### ISSUE TITLE
Working around intermittent 501 from Confluence 7.x?

### ISSUE THREAD

**Issue by nrbgt** [sphinx-contrib/confluencebuilder#314]
This seems to have cropped up recently:

https://jira.atlassian.com/browse/CONFSERVER-59196

While there is no API for the work-around, doing a "normal" page view does appear to clear it up.  I've done a little monkeypatch to get our pipeline up and running:

```py
import os
import re

import requests
from sphinxcontrib.confluencebuilder.builder import ConfluenceBuilder
from sphinxcontrib.confluencebuilder.exceptions import ConfluenceBadApiError

re_page_id = r"Unable to save changes to unreconciled page ContentId\{id=(\d+)\}"


def patch_refresh():
    _orig_publish_doc = ConfluenceBuilder.publish_doc

    def publish_doc(self, docname, output):
        try:
            return _orig_publish_doc(self, docname, output)
        except ConfluenceBadApiError as err:
            err_str = str(err)
            if "501" in err_str:
                print("{}: HTTP 501, manually fetching page...".format(docname))
                page_id = re.findall(re_page_id, str(err))
                if page_id:
                    requests.get(
                        "{}pages/viewpage.action?pageId={}".format(
                            os.environ["CFX_URL"], page_id[0]
                        ),
                        auth=(os.environ["CFX_AUTH_USR"], os.environ["CFX_AUTH_PSW"]),
                    )
                print("{}: trying again...".format(docname))
                return _orig_publish_doc(self, docname, output)
            raise err

    ConfluenceBuilder.publish_doc = publish_doc
    print("Patched ConfluenceBuilder.publish_doc!")
```

Is there any interest in a PR for this, or would it be more prudent to await an "official" solution?

**Comment by jdknight** [comment]
@nrbgt, thanks for mentioning this and referencing this JIRA issue -- an interesting read :expressionless:.

While it _appears_ to be fixed in Confluence v7.3.3, it would be nice to add some support in this extension for users who are using the issues versions of Confluence. If you would like to submit a PR for this, I would be more than willing to help review and re-test it on instances (I may setup some of these issued instances myself to see the error in action). If not, I can try to get around to this later (after addressing #312).

I assume from the provided implementation, the goal is to (to help my understanding):

1. Detect a "unreconciled page" error and extract the identifier for the page being updated.
2. Manually perform a `GET` request to the page to help Confluence in its synchronization.
3. Re-attempt to publish the page.

(neat hack 👍 )

**Comment by jdknight** [comment]
Was taking a look at this and setting up instances to reproduce for testing a near-matching hotfix. I only then realized that the page refresh interaction can most likely result in a required authentication outside of an (authenticated) API access point. Since this brings up addition complexity for a small use case (and hopefully now rarer use case since there have been ~5 minor version bumps which include a fix for this), I am now opting to not introduce a workaround for this Confluence instance issue.

There was a change introduced in #375 which should help notify a user in the event that they do detect an unreconciled error scenario, which now will provide a more detailed error message to help a user continue. Hopefully that will be sufficient enough until the pre-v7.3.3. series has been dropped for support.

Will mark this issue as closed; the suggestion was appreciated.
